### PR TITLE
TIP-706-family: Add a Family Filter for Elasticsearch

### PR DESCRIPTION
**Pull request by anaelChardan**
**Verify raw mapping**

**ES Filter Checklist:**
- [x] Review the PQB filter integration tests
- [x] Add index mapping integration tests (not needed)
- [x] Write the filter & Spec
- [x] Update the documentation in `search_specification.rst`
- [x] Is the PIM installable ?

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst [hunk 1]
```diff
@@ -898,16 +898,13 @@ Same as ``IN``, but with the ``must_not`` occured type
 
 Family
 ******
-:Apply: "family" field
+:Apply: apply on 'family' field
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: on the

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst [hunk 2]
```diff
@@ -917,10 +914,37 @@ IN
...
 
-    terms:
-        family.id: [5, 6 7]
+    'terms' => [
+        'family' => ['familyA', 'familyB', 'familyC']
+    ]
+
+NOT IN
+~~~~~~
+:Type: mustNot
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: must_not

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst [hunk 3]
```diff
@@ -917,10 +914,37 @@ IN
...
+        'family' => ['familyA', 'familyB', 'familyC']
+    ]
+
+IS EMPTY
+~~~~~~~~
+:Type: mustNot
+
+.. code-block:: php
+
+    'exists' => ['field' => 'family']
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: can you add wrapping [ please ?

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst (hunk 1 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: Ok, I correct, can I correct for the category too ? -> :Apply: apply on 'categories' field

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst (hunk 3 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: Yup

File: src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst (hunk 2 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/tests/integration/ElasticSearch/Filter/search_specification.rst]: Ok

**Review by SamirBoulil — APPROVED** [review]
Really small comments. Good job dude ! Keep on rocking! 👍  #Depiling

Don't forget to check the PIM installs and the CI without behats.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        $value,
+        $locale = null,
+        $scope = null,
+        $options = []
+    ) {
+        if (null === $this->searchQueryBuilder) {
+            throw new \LogicException('The search query builder is not initialized in the filter.');
+        }
+
+        if (Operators::IS_EMPTY == $operator || Operators::IS_NOT_EMPTY) {
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php]: `Operators::IS_EMPTY === $operator || Operators::IS_NOT_EMPTY === $operator`

it's weird your tests are green, maybe some of them are missing with the `Operators::IS_NOT_EMPTY` operator

**Review by skeleton — APPROVED** [review]

**Review by damien-carcel — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php (hunk 4 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php]: Indeed, it's always true ^^.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        if (Operators::IS_EMPTY !== $operator && Operators::IS_NOT_EMPTY !== $operator) {
+            $this->checkValue($field, $value);
+        }
+
+        switch ($operator) {
+            case Operators::IN_LIST:
+                $clause = [
+                    'terms' => [
```
Comment (momoss) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php:55]: Is it not 'term' ?

Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php:55]: No, it's terms -> https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-terms-query.html

Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/FamilyFilter.php:55]: term is for exact, and terms is like a "in" clause :-) -> The terms query is also aliased with in as the filter name for simpler usage [5.0.0] Deprecated in 5.0.0. use terms instead .

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    function it_is_initializable() {
+        $this->shouldHaveType(FamilyFilter::class);
+    }
+
+    function it_is_a_filter()
+    {
+        $this->shouldImplement(FieldFilterInterface::class);
+        $this->shouldBeAnInstanceOf(AbstractFieldFilter::class);
```
Comment (arnolanglade) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:37]: Same feedback than the previous PR ;)

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    function it_adds_a_filter_with_operator_in_list(SearchQueryBuilder $sqb) {
+        $sqb->addFilter(
+            [
+                'terms' => [
+                    'family' => ['familyA']
+                ]
+            ]
+        )->shouldBeCalled();
+
+        $this->setQueryBuilder($sqb);
```
Comment (arnolanglade) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:66]: You could call this method in `let`.

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        $sqb->addMustNot(
+            [
+                'terms' => [
+                    'family' => ['familyA']
+                ]
+            ]
+        )->shouldBeCalled();
+
+        $this->setQueryBuilder($sqb);
+        $this->addFieldFilter('family', Operators::NOT_IN_LIST, ['familyA'], null, null, []);
```
Comment (arnolanglade) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:80]: What happens if the fourth and fifth parameters are not empties?
The last parameter is an array, perhaps method "options"? Which are the allowed options?

**Review by arnolanglade — COMMENTED** [review]
Still some little comments but the rest seems good!

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php (hunk 8 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:80]: We don't know it seems that it 's not used :/

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php (hunk 7 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:66]: +1

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php (hunk 6 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:37]: Yup

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php (hunk 8 above)
Comment (arnolanglade) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:80]: yes that means it miss examples to describe that those parameters are not used.

Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/FamilyFilterSpec.php:80]: We'll investigate with momoss :)

File: src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml [hunk 9]
```diff
@@ -1,12 +1,13 @@
 parameters:
-    pim_catalog.query.product_query_builder_factory.class:      Pim\Component\Catalog\Query\ProductQueryBuilderFactory
-    pim_catalog.query.product_query_builder.class:              Pim\Component\Catalog\Query\ProductQueryBuilder
-    pim_catalog.query.filter.registry.class:                    Pim\Component\Catalog\Query\Filter\FilterRegistry
-    pim_catalog.query.sorter.registry.class:                    Pim\Component\Catalog\Query\Sorter\SorterRegistry
-    pim_catalog.query.filter.field_dumper.class:                Pim\Bundle\CatalogBundle\Command\ProductQueryHelp\FieldFilterDumper
-    pim_catalog.query.filter.attribute_dumper.class:            Pim\Bundle\CatalogBundle\Command\ProductQueryHelp\AttributeFilterDumper
-    pim_catalog.doctrine.query.filter.object_id_resolver.class: Pim\Bundle\CatalogBundle\Doctrine\Common\Filter\ObjectIdResolver
+    pim_catalog.query.product_query_builder_factory.class:        Pim\Component\Catalog\Query\ProductQueryBuilderFactory
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml:2]: don't do things like this please, 
it breaks the history and the readibility of the PR

**Review by jjanvier — APPROVED** [review]
nice! GTM once spec fixed

**Review by SamirBoulil — APPROVED** [review]

**Pull request merged**